common/src/test/scala/tsec/SymmetricSpec.scala

```diff
@@ -0,0 +1,278 @@
+        val testEncryptionDecryption: IO[String] = for {
+          key       <- S.generateKey
+          encrypted <- E.encryptWithAAD(testPlainText, key, aad)
+          decrypted <- E.decryptWithAAD(encrypted, key, aad)
+        } yield decrypted.toUtf8String
+        if (!testMessage.isEmpty)
+          testEncryptionDecryption.attempt.unsafeRunSync() must equal(Right(testMessage))
+      }
+    }
+
+    it should "not decrypt properly for an incorrect key" in {
+      forAll { (testMessage: String) =>
+        val testPlainText = PlainText(testMessage.utf8Bytes)
+        val testEncryptionDecryption: IO[String] = for {
+          key1      <- S.generateKey
+          key2      <- S.generateKey
+          encrypted <- E.encrypt(testPlainText, key1)
+          decrypted <- E.decrypt(encrypted, key2)
+        } yield decrypted.toUtf8String
+        if (!testMessage.isEmpty)
+          testEncryptionDecryption.attempt.unsafeRunSync() mustNot equal(Right(testMessage))
+      }
+    }
+
+    it should "not decrypt properly for correct key but incorrect AAD" in {
+      forAll { (testMessage: String, AAD1: String, AAD2: String) =>
+        val testPlainText = PlainText(testMessage.utf8Bytes)
+        val aad1          = AAD(AAD1.utf8Bytes)
+        val aad2          = AAD(AAD2.utf8Bytes)
+        val testEncryptionDecryption: IO[String] = for {
+          key1      <- S.generateKey
+          encrypted <- E.encryptWithAAD(testPlainText, key1, aad1)
+          decrypted <- E.decryptWithAAD(encrypted, key1, aad2)
+        } yield decrypted.toUtf8String
+        if (!testMessage.isEmpty && !AAD1.isEmpty && !AAD2.isEmpty)
+          testEncryptionDecryption.attempt.unsafeRunSync() mustNot equal(Right(testMessage))
+      }
+    }
+
+    /** Detached mode tests **/
+    it should "Encrypt and decrypt for the same key in detached mode" in {
+      forAll { (testMessage: String) =>
+        val testPlainText = PlainText(testMessage.utf8Bytes)
+        val testEncryptionDecryption: IO[String] = for {
+          key       <- S.generateKey
+          encrypted <- E.encryptDetached(testPlainText, key)
+          decrypted <- E.decryptDetached(encrypted._1, key, encrypted._2)
+        } yield decrypted.toUtf8String
+        if (!testMessage.isEmpty)
+          testEncryptionDecryption.attempt.unsafeRunSync() must equal(Right(testMessage))
+      }
+    }
+
+    it should "Encrypt and decrypt for the same key and AEAD in detached mode" in {
+      forAll { (testMessage: String, aadData: String) =>
+        val testPlainText = PlainText(testMessage.utf8Bytes)
+        val aad           = AAD(aadData.utf8Bytes)
+        val testEncryptionDecryption: IO[String] = for {
+          key       <- S.generateKey
+          encrypted <- E.encryptWithAADDetached(testPlainText, key, aad)
+          decrypted <- E.decryptWithAADDetached(encrypted._1, key, aad, encrypted._2)
+        } yield decrypted.toUtf8String
+        if (!testMessage.isEmpty)
```

Comment:
maybe we should use _whenever_

Reply:
tbh, I'm not a fan of using `scalatest` syntax, mostly because I don't want to be married to it.

To be quite honest, I do not enjoy scalatest in the slightest, nor specs2, but we live with it because there's nothing else (Yet at least. Edmund hyped me up a little bit ago about a testing library he's working on).